src/magic_folder/util/database.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2020 The Magic-Folder Developers
+# See COPYING for details.
+
+"""
+Utilties for dealing with sqlite.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import contextlib
+import inspect
+import sqlite3
+
+import attr
+from six import wraps
+from twisted.python.compat import currentframe
+
+__all__ = [
+    "LockableDatabase",
+    "RecusiveTransaction",
+    "WithCursorGenerator",
+    "with_cursor",
+]
+
+
+@attr.s(auto_exc=True, frozen=True)
+class WithCursorGenerator(TypeError):
+    """
+    :py:`with_cursor` cannot be used on a generator.
+
+    As it is currently implemented, the transaction would be commited before
+    any code in the decorated generator ran. Even if we special cased generators,
+    having the transaction stay open until the generator completes, that loses
+    the proprerty of `with_cursor` being transparent and atomic from the point
+    of view of the calller.
+    """
+
+    function = attr.ib()
+
+    def __str__(self):
+        return "'with_cursor' cannot decorate a generator function: {!r}".format(
+            self.function
+        )
+
+
+@attr.s(auto_exc=True, frozen=True)
+class RecusiveTransaction(Exception):
+    function = attr.ib(default=None)
+    caller = attr.ib(default=None)
+
+    def __str__(self):
+        message = "Tried to enter a recursive transaction"
+        if self.function:
+            message += " when calling '{}'".format(self.function)
+        if self.caller:
+            message += " from '{}'".format(self.caller)
+        message += "."
+        return message
+
+
+@attr.s
+class LockableDatabase(object):
+    """
+    Wrapper around an :py:`sqlite3.Connection`
+    """
+
+    _database = attr.ib(validator=attr.validators.instance_of(sqlite3.Connection))
+    _in_transaction = attr.ib(
+        default=False, init=False, validator=attr.validators.instance_of(bool)
+    )
+
+    @contextlib.contextmanager
+    def transaction(self, function_name=None):
+        """
+        Context manager that starts a transaction and rollsback
+        if an exception is raised.
+
+        :param Optional[function_name]: The name of the calling function, used
```

Comment:
`:param Optional[str] function_name:` perhaps?  However, there seems to be exactly one caller and it supplies a value.  So maybe it needn't be optional?  Unless `f.__name__` is allowed to be `None`...

Reply:
I do want to support direct usage of this function as well. I think, for example, that `GlobalConfigDatabase` may want to use this in some places.